crates/turbopack-ecmascript-plugins/src/transform/swc_ecma_transform_plugins.rs

```diff
@@ -0,0 +1,223 @@
+/// it to operate with the turbo_task caching requirements.
+/// Internally this contains a `CompiledPluginModuleBytes`, which points to the
+/// compiled, serialized wasmer::Module instead of raw file bytes to reduce the
+/// cost of the compilation.
+#[turbo_tasks::value(
+    transparent,
+    serialization = "none",
+    eq = "manual",
+    into = "new",
+    cell = "new"
+)]
+pub struct SwcPluginModule(
+    #[turbo_tasks(trace_ignore)]
+    #[cfg(feature = "swc_ecma_transform_plugin")]
+    pub swc_core::plugin_runner::plugin_module_bytes::CompiledPluginModuleBytes,
+    // Dummy field to avoid turbo_tasks macro complains about empty struct.
+    // This is due to we can't import CompiledPluginModuleBytes by default, it should be only
+    // available for the target / platforms can support swc plugins (which can build wasmer)
+    #[cfg(not(feature = "swc_ecma_transform_plugin"))] pub (),
+);
+
+impl SwcPluginModule {
+    pub fn new(plugin_name: &str, plugin_bytes: Vec<u8>) -> Self {
+        Self {
+            #[cfg(feature = "swc_ecma_transform_plugin")]
+            0: {
+                use swc_core::plugin_runner::plugin_module_bytes::{
+                    CompiledPluginModuleBytes, RawPluginModuleBytes,
+                };
+                CompiledPluginModuleBytes::from(RawPluginModuleBytes::new(
+                    plugin_name.to_string(),
+                    plugin_bytes,
+                ))
+            },
+            #[cfg(not(feature = "swc_ecma_transform_plugin"))]
+            0: (),
+        }
+    }
+}
+
+#[turbo_tasks::value(shared)]
+struct UnsupportedSwcEcmaTransformPluginsIssue {
+    pub context: FileSystemPathVc,
+}
+
+#[turbo_tasks::value_impl]
+impl Issue for UnsupportedSwcEcmaTransformPluginsIssue {
+    #[turbo_tasks::function]
+    fn severity(&self) -> IssueSeverityVc {
+        IssueSeverity::Warning.into()
+    }
+
+    #[turbo_tasks::function]
+    fn category(&self) -> StringVc {
+        StringVc::cell("transform".to_string())
+    }
+
+    #[turbo_tasks::function]
+    async fn title(&self) -> Result<StringVc> {
+        Ok(StringVc::cell(format!(
+            "Unsupported SWC Ecma transform plugins on this platform."
+        )))
+    }
+
+    #[turbo_tasks::function]
+    fn context(&self) -> FileSystemPathVc {
+        self.context
+    }
+
+    #[turbo_tasks::function]
+    fn description(&self) -> StringVc {
+        StringVc::cell(
+            "Turbopack does not yet support running SWC ecma transform plugins on this platform."
+                .to_string(),
+        )
+    }
+}
+
+#[derive(Debug)]
+pub struct SwcEcmaTransformPluginsTransformer {
+    #[cfg(feature = "swc_ecma_transform_plugin")]
+    plugins: Vec<(SwcPluginModuleVc, serde_json::Value)>,
+}
+
+impl SwcEcmaTransformPluginsTransformer {
+    #[cfg(feature = "swc_ecma_transform_plugin")]
+    pub fn new(plugins: Vec<(SwcPluginModuleVc, serde_json::Value)>) -> Self {
```

Comment:
To make the API clearer, this should either accept:
```suggestion
    pub fn new(plugins: Vec<SwcEcmaTransformPlugin>) -> Self {
```
where
```rust
pub struct SwcEcmaTransformPlugin {
    module: SwcPluginModuleVc,
    config: serde_json::Value
}
```

or

```suggestion
    pub fn new(plugins: Vec<(SwcPluginModuleVc, SwcPluginConfig)>) -> Self {
```

where `SwcPluginConfig` is a newtype:

```rust
pub struct SwcPluginConfig(serde_json::Value);
```

Reply:
This is sort of intentional: currently upstream swc's config itself is not finalized for the plugins, so I'd rather take it out separately for now and then visit everything at once with changes. It'll be breaking changes, but swc's plugin is experimental anyway.